_drafts/2024-09-01-aws-terraform-02.md

```diff
@@ -0,0 +1,267 @@
+
+The `image_tag_mutability` property controls whether tags can be reused.  
+
+When set to `MUTABLE`, tags can be reused, allowing clients to pull the latest image automatically (e.g. via the `latest` tag).  
+
+When set to `IMMUTABLE` tags are fixed to specific images which prevents automatic updates but provides clients with greater stability and control as the image wont change unless the tag itself is updated.
+
+The `scan_on_push` property controls whether images should be scanned for vulnerabilities upon being pushed to the repository. As it's free to enable, we might as well do so in case it finds any [CVEs](https://cve.mitre.org/).
+
+The `name` property should be self-evident.
+
+Let's run `terraform apply` to create this repository, and make note of the repository uri output.
+
+## Step 2 - Building our web server image and pushing it to ECR
+
+Now that we've defined an ECR repo, we need a container image to host within the repo. This image could be whatever we want it to be, but in our case, let's make it a simple node web api.  
+
+Here's one I prepared earlier! Clone [this repo](https://github.com/kstolen0/sample-node-project) and then navigate to the directory.
+
+Create a new bash script named `build-and-push.sh` and add the following:
+
+```bash
+#!/bin/bash
+
+aws --profile [your-aws-cli-profile] ecr get-login-password --region [region] | sudo docker login --username AWS --password-stdin [ecr-uri]
+sudo docker build -t unicorn-api .
+sudo docker tag unicorn-api:latest [ecr-uri]:latest
+sudo docker push [ecr-uri]:latest
+aws --profile [your-aws-cli-profile] ecr list-images --repository-name unicorn-api
+
+```
+We'll also need to enable read/write/execute permissions on the script in order to run it.
+
+`chmod 700 build-and-push.sh`  
+
+Now run the script and if all went well you should see the following output:
+
+
+![console output from pushing an image container to ecr](/assets/getting-started-with-aws-tf-02/example-push-ecr-output.png)
+
+
+## Step 3 - Pull the ECR image onto EC2 instance and run it
+
+By now we should have an EC2 instance running in AWS and a container image hosted on ECR. Now we just need to run the image on our EC2 instance.
+
+In order to achieve this we'll need to:
+1. Add a security egress rule to allow EC2 to connect to ECR
+2. Define a role to allow read access to ECR and assign it to our EC2 instance
+3. write another custom script to run on our instance which will pull the docker image and run it in its own Docker environment 
+
+Lets define our security group rule first.
+
+Open the `unicorn-api.tf` file and add the following egress rule:
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_vpc_security_group_egress_rule" "allow_outgoing_https" {
+  
+  security_group_id = aws_security_group.unicorn_api_security_group.id
+
+  description = "allow egress via port 443"
+  cidr_ipv4   = "0.0.0.0/0"
+  ip_protocol = "tcp"
+  from_port   = 443
+  to_port     = 443
+}
+```
```

Comment:
This is what I have in my reference project, but I thought there was an alternative to just specify aws services..